app/models/user.rb

```diff
@@ -162,8 +164,50 @@ def remember_me?
     remember_token_expires_at && remember_token_expires_at > Time.zone.now
   end
 
+  def mfa_enabled?
+    !no_mfa?
+  end
+
+  def disable_mfa!
+    no_mfa!
+    self.mfa_seed = ''
+    self.mfa_recovery_codes = []
+    self.last_otp_at = nil
+    save!(validate: false)
+  end
+
+  def enable_mfa!(seed, level)
+    self.mfa_level = level
+    self.mfa_seed = seed
+    self.mfa_recovery_codes = Array.new(10).map { SecureRandom.hex(6) }
+    save!(validate: false)
+  end
+
+  def otp_verified?(otp)
+    if no_mfa? || verify_digit_otp(otp)
+      true
+    else
+      is_recovery = mfa_recovery_codes.include?(otp)
+      if is_recovery
+        mfa_recovery_codes.delete(otp)
+        save!(validate: false)
+      end
```

Comment:
please change this to `return false unless is_recovery`. last line in `otp_verified?` could be just `save!(validate: false)`